packages/playground/website/src/lib/router-hooks.ts

```diff
@@ -0,0 +1,103 @@
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import { useSearch, useLocation } from 'wouter';
+
+export function useSearchParams() {
+	const search = useSearch();
+	const location = useCurrentUrl();
+	const [, setLocation] = useLocation();
+	return [
+		useMemo(
+			() => new URL('?' + search, window.location.href).searchParams,
+			[search]
+		),
+		useCallback(
+			(params: Record<string, string>) => {
+				const currentUrl = new URL(location);
+				Object.entries(params).forEach(([key, value]) => {
+					currentUrl.searchParams.set(key, value);
+				});
+				setLocation(currentUrl.toString());
+			},
+			[setLocation, location]
+		),
+	] as const;
+}
+
+/**
+ * Wouter's useLocation hook doesn't reflect the search params or the fragment.
+ * This hook does.
+ *
+ * The code below is copied from wouter's useLocation hook.
+ * @see https://github.com/molefrog/wouter
+ */
+
+/**
+ * History API docs @see https://developer.mozilla.org/en-US/docs/Web/API/History
+ */
+const eventPopstate = 'popstate';
+const eventPushState = 'pushState';
+const eventReplaceState = 'replaceState';
+const eventHashchange = 'hashchange';
+const events = [
+	eventPopstate,
+	eventPushState,
+	eventReplaceState,
+	eventHashchange,
+];
+
+const subscribeToLocationUpdates = (callback: () => void) => {
+	for (const event of events) {
+		window.addEventListener(event, callback);
+	}
+	return () => {
+		for (const event of events) {
+			window.removeEventListener(event, callback);
+		}
+	};
+};
+
+const patchKey = Symbol.for('wouter_v3');
+
+// While History API does have `popstate` event, the only
+// proper way to listen to changes via `push/replaceState`
+// is to monkey-patch these methods.
```

Comment:
If we're using something that monkey-patches browser APIs, it's helpful happenstance that we ended up exposing that fact in our own codebase. It's good to know about because if we ever ended up debugging something in this area, it could be hard to notice the monkey patch.